sourcecode/hub/app/Http/Controllers/GoogleController.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Models\User;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Hash;
+use Laravel\Socialite\Facades\Socialite;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+
+class GoogleController extends Controller
+{
+    public function loginWithGoogle(): RedirectResponse
+    {
+        return Socialite::driver('google')->redirect();
+    }
+
+    public function callbackFromGoogle(): RedirectResponse
+    {
+        try {
+            $user = Socialite::driver('google')->user();
+            $existingUser = User::where('email', $user->getEmail())->first();
```

Comment:
This variable and the subsequent `if/else` can be replaced with a simple `User::updateOrCreate()`.

https://laravel.com/docs/10.x/eloquent#upserts